add _proxyName/_isContentFromNetwork to WKWebView

nmahendru opened the pull request:
#### ee22bac374c677ae4426ba1775556821eb44410a
<pre>
add _proxyName/_isContentFromNetwork to WKWebView
<a href="https://rdar.apple.com/144167580">rdar://144167580</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=287023">https://bugs.webkit.org/show_bug.cgi?id=287023</a>

Reviewed by Alex Christensen.

_proxyName/_isContentFromNetwork will only have a useful value when the pageload is committed
for the main frame. In other words, only use these values when you get the callback:
(void)_webView:(WKWebView *)webView didCommitLoadWithRequest:(NSURLRequest *)request inFrame:(WKFrameInfo *)frame

_isContentFromNetwork is useful differentiate between whether the response is from a cache or network.
If it&apos;s coming from a cache, then _wasPrivateRelayed/_proxyName might not have the correct values
as it is possible the page was cached in absence of icloud+.
Also these values are only useful for mainframe navigation. So they should only be checked:
[webView _mainFrame] != nil

* Source/WebCore/platform/network/ResourceResponseBase.h:
(WebCore::ResourceResponseData::ResourceResponseData):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _proxyName]):
(-[WKWebView _isContentFromNetwork]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/PageLoadState.cpp:
(WebKit::PageLoadState::didCommitLoad):
* Source/WebKit/UIProcess/PageLoadState.h:
(WebKit::PageLoadState::proxyName):
(WebKit::PageLoadState::source):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::commitProvisionalFrame):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::commitProvisionalPage):
(WebKit::WebPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidCommitLoad):

Canonical link: <a href="https://commits.webkit.org/289937@main">https://commits.webkit.org/289937@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff5848e67820f6cc7af0061128573cf6fb40a4fe

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88461 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7980 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42901 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93419 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39215 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90512 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8367 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16164 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68236 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25952 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91463 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6407 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80015 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48602 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/87960 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6179 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/34430 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38323 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/76552 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35314 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95261 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15636 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11459 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77098 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75871 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76358 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20746 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19164 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8675 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13825 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15652 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20960 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15393 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18842 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17175 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -163,6 +163,9 @@ typedef NS_ENUM(NSInteger, _WKImmediateActionType) {
 
 @interface WKWebView (WKPrivate)
 
+@property (nonatomic, readonly) NSString *_proxyName;
```
Needs WK_API_AVAILABLE.

nmahendru replied:
absolutely. done.

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
```diff
@@ -163,6 +163,9 @@ typedef NS_ENUM(NSInteger, _WKImmediateActionType) {
 @interface WKWebView (WKPrivate)
 
+@property (nonatomic, readonly) NSString *_proxyName;
+@property (nonatomic, readonly) BOOL _isPageLoadCommitted;
```
I don't think we'll need this.

nmahendru replied:
This let's the client check if the call back has already been called and in case they want to use the other values somewhere other than the callback. The client could also cache the fact that they have received a callback at their end but this is just another way.

achristensen07 commented on Source/WebKit/UIProcess/PageLoadState.h:
```diff
@@ -152,6 +159,8 @@ class PageLoadState {
     void negotiatedLegacyTLS(const Transaction::Token&);
     bool wasPrivateRelayed() const { return m_committedState.wasPrivateRelayed; }
+    String proxyName() { return m_committedState.proxyName; }
+    WebCore::ResourceResponseBaseSource source() { return m_committedState.source; }
```
We should probably rename this enum to remove "Base" from its name.  That is a confusing implementation detail of how we make ResourceResponse implemented on different platforms.

nmahendru replied:
Good idea. Done.

achristensen07 commented on Source/WebKit/UIProcess/RemotePageProxy.h:
```diff
@@ -104,7 +104,7 @@ class RemotePageProxy : public IPC::MessageReceiver, public RefCounted<RemotePag
     bool didReceiveSyncMessage(IPC::Connection&, IPC::Decoder&, UniqueRef<IPC::Encoder>&) final;
     void decidePolicyForResponse(FrameInfoData&&, std::optional<WebCore::NavigationIdentifier>, const WebCore::ResourceResponse&, const WebCore::ResourceRequest&, bool canShowMIMEType, const String& downloadAttribute, bool isShowingInitialAboutBlank, WebCore::CrossOriginOpenerPolicyValue activeDocumentCOOPValue, CompletionHandler<void(PolicyDecision&&)>&&);
-    void didCommitLoadForFrame(IPC::Connection&, WebCore::FrameIdentifier, FrameInfoData&&, WebCore::ResourceRequest&&, std::optional<WebCore::NavigationIdentifier>, const String& mimeType, bool frameHasCustomContentProvider, WebCore::FrameLoadType, const WebCore::CertificateInfo&, bool usedLegacyTLS, bool privateRelayed, bool containsPluginDocument, WebCore::HasInsecureContent, WebCore::MouseEve ...
+    void didCommitLoadForFrame(IPC::Connection&, WebCore::FrameIdentifier, FrameInfoData&&, WebCore::ResourceRequest&&, std::optional<WebCore::NavigationIdentifier>, const String& mimeType, bool frameHasCustomContentProvider, WebCore::FrameLoadType, const WebCore::CertificateInfo&, bool usedLegacyTLS, bool privateRelayed, const String& proxyName, WebCore::ResourceResponseBaseSource, bool contains ...
```
Can we put usedLegacyTLS, privateRelayed, proxyName, and the source into a struct named something like "ConnectionInfo"?  That would make this less messy.

nmahendru replied:
Will fast follow on a change for putting all the params in a struct.

achristensen07 commented on Source/WebKit/UIProcess/WebFrameProxy.cpp:
```diff
@@ -449,7 +449,7 @@ void WebFrameProxy::commitProvisionalFrame(IPC::Connection& connection, FrameIde
         }
     }
-    protectedPage()->didCommitLoadForFrame(connection, frameID, WTFMove(frameInfo), WTFMove(request), navigationID, mimeType, frameHasCustomContentProvider, frameLoadType, certificateInfo, usedLegacyTLS, privateRelayed, containsPluginDocument, hasInsecureContent, mouseEventPolicy, userData);
+    protectedPage()->didCommitLoadForFrame(connection, frameID, WTFMove(frameInfo), WTFMove(request), navigationID, mimeType, frameHasCustomContentProvider, frameLoadType, certificateInfo, usedLegacyTLS, privateRelayed, proxyName, WebCore::ResourceResponseBaseSource::Unknown, containsPluginDocument, hasInsecureContent, mouseEventPolicy, userData);
```
The sender of Messages::WebPageProxy::DidCommitLoadForFrame can get the source to here instead of using ::Unknown.

nmahendru replied:
done.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4978,7 +4978,7 @@ void WebPageProxy::commitProvisionalPage(IPC::Connection& connection, FrameIdent
     swapToProvisionalPage(provisionalPage.releaseNonNull());
 
-    didCommitLoadForFrame(connection, frameID, WTFMove(frameInfo), WTFMove(request), navigationID, mimeType, frameHasCustomContentProvider, frameLoadType, certificateInfo, usedLegacyTLS, privateRelayed, containsPluginDocument, hasInsecureContent, mouseEventPolicy, userData);
+    didCommitLoadForFrame(connection, frameID, WTFMove(frameInfo), WTFMove(request), navigationID, mimeType, frameHasCustomContentProvider, frameLoadType, certificateInfo, usedLegacyTLS, privateRelayed, String { }, WebCore::ResourceResponseBaseSource::Unknown, containsPluginDocument, hasInsecureContent, mouseEventPolicy, userData);
```
Ditto about use of unknown.  Same with the proxy name.  That'll be clearer when they're all in a struct.

nmahendru replied:
done.

achristensen07 commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -3537,7 +3537,7 @@ enum class WebCore::ResourceLoadPriority : uint8_t {
 };
 
-[Nested] enum class WebCore::ResourceResponseBaseSource : uint8_t {
+[CustomHeader] enum class WebCore::ResourceResponseBaseSource : uint8_t {
```
If we put this in a struct, it can remain Nested.

nmahendru replied:
Will do that asa follow on

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -2922,6 +2922,16 @@ - (BOOL)_wasPrivateRelayed
+    return _page->pageLoadState().proxyName();
+}
+
+- (BOOL)_isFromNetwork
```
_contentIsFromNetwork?
The WKWebView isn't from the network.  The content is.

achristensen07 commented on Source/WebKit/UIProcess/PageLoadState.cpp:
```diff
@@ -134,6 +135,8 @@ void PageLoadState::commitChanges()
     if (wasPrivateRelayedChanged)
         callObserverCallback(&Observer::willChangeWasPrivateRelayed);
+    if (proxyNameChanged)
+        callObserverCallback(&Observer::willProxyNameChanged);
```
This doesn't do anything.  Remove.

achristensen07 commented on Source/WebKit/UIProcess/PageLoadState.h:
```diff
@@ -59,6 +63,9 @@ class PageLoadStateObserverBase : public AbstractRefCountedAndCanMakeWeakPtr<Pag
     virtual void willChangeWasPrivateRelayed() { }
     virtual void didChangeWasPrivateRelayed() { }
 
+    virtual void willProxyNameChanged() { }
```
These have no implementation that does anything.  Remove.

achristensen07 reviewed (approved):
It looks like you got halfway to making one of your two new properties KVO observable.  Let's remove that halfway until we get a request to make both KVO observable.

nmahendru committed:
add _proxyName/_isContentFromNetwork to WKWebView
rdar://144167580
https://bugs.webkit.org/show_bug.cgi?id=287023

Reviewed by Alex Christensen.

_proxyName/_isContentFromNetwork will only have a useful value when the pageload is committed
for the main frame. In other words, only use these values when you get the callback:
(void)_webView:(WKWebView *)webView didCommitLoadWithRequest:(NSURLRequest *)request inFrame:(WKFrameInfo *)frame

_isContentFromNetwork is useful differentiate between whether the response is from a cache or network.
If it's coming from a cache, then _wasPrivateRelayed/_proxyName might not have the correct values
as it is possible the page was cached in absence of icloud+.
Also these values are only useful for mainframe navigation. So they should only be checked:
[webView _mainFrame] != nil

* Source/WebCore/platform/network/ResourceResponseBase.h:
(WebCore::ResourceResponseData::ResourceResponseData):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _proxyName]):
(-[WKWebView _isContentFromNetwork]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/PageLoadState.cpp:
(WebKit::PageLoadState::didCommitLoad):
* Source/WebKit/UIProcess/PageLoadState.h:
(WebKit::PageLoadState::proxyName):
(WebKit::PageLoadState::source):
* Source/WebKit/UIProcess/ProvisionalPageProxy.cpp:
(WebKit::ProvisionalPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/ProvisionalPageProxy.h:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
(WebKit::WebFrameProxy::commitProvisionalFrame):
* Source/WebKit/UIProcess/WebFrameProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::commitProvisionalPage):
(WebKit::WebPageProxy::didCommitLoadForFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidCommitLoad):

Canonical link: https://commits.webkit.org/289937@main
